elasticdl/python/master/k8s_instance_manager.py

```diff
@@ -218,10 +220,18 @@ def _event_cb(self, event):
         worker_id = -1
         ps_id = -1
         with self._lock:
+            if pod_name in self._failed_pods:
+                return
             if pod_name in self._worker_pod_name_to_id:
                 worker_id = self._worker_pod_name_to_id.get(pod_name)
                 self._worker_pods_phase[worker_id] = (pod_name, phase)
-                if evt_type == "DELETED":
+                # Workaround for memory leak issues in tf eager mode.
+                # A pod may fail due to OOM from tf eager mode memory leak.
+                failed_pod = False
+                if evt_type == "MODIFIED" and phase == "Failed":
+                    self._failed_pods.append(pod_name)
+                    failed_pod = True
+                if evt_type == "DELETED" or failed_pod:
```

Comment:
Need we add a counter for each failed pod? In order to avoid the infinite recover for other causes.

Reply:
When a pod fails, it is moved into `self._failed_pods.append` and a new pod will be launched to substitute it with a different pod name (with the same service). Each pod will only fail once.

Also, there is no way to know the cause of failure from pod status.